runtime/compiler/il/J9MethodSymbol.cpp

```diff
@@ -450,6 +450,97 @@ J9::MethodSymbol::safeToSkipArrayStoreChecks()
    return false;
    }
 
+static TR::RecognizedMethod canSkipNonNullableArrayNullStoreCheck[] = {
+   TR::java_lang_invoke_CollectHandle_invokeExact,
+   TR::java_util_ArrayList_add,
+   TR::java_util_ArrayList_ensureCapacity,
+   TR::java_util_ArrayList_get,
+   TR::java_util_ArrayList_remove,
+   TR::java_util_ArrayList_set,
+   TR::java_util_Hashtable_get,
+   TR::java_util_Hashtable_put,
+   TR::java_util_Hashtable_clone,
+   TR::java_util_Hashtable_putAll,
+   TR::java_util_Hashtable_rehash,
+   TR::java_util_Hashtable_remove,
+   TR::java_util_Hashtable_contains,
+   TR::java_util_Hashtable_getEntry,
+   TR::java_util_Hashtable_getEnumeration,
+   TR::java_util_Hashtable_elements,
+   TR::java_util_HashtableHashEnumerator_hasMoreElements,
+   TR::java_util_HashtableHashEnumerator_nextElement,
+   TR::java_util_HashMap_rehash,
+   TR::java_util_HashMap_analyzeMap,
+   TR::java_util_HashMap_calculateCapacity,
+   TR::java_util_HashMap_findNullKeyEntry,
+   TR::java_util_HashMap_get,
+   TR::java_util_HashMap_getNode,
+   TR::java_util_HashMap_putImpl,
+   TR::java_util_HashMap_findNonNullKeyEntry,
+   TR::java_util_HashMap_resize,
+   TR::java_util_HashMap_prepareArray,
+   TR::java_util_HashMap_keysToArray,
+   TR::java_util_HashMap_valuesToArray,
+   TR::java_util_HashMapHashIterator_nextNode,
+   TR::java_util_HashMapHashIterator_init,
+   TR::java_util_Vector_addElement,
+   TR::java_util_Vector_contains,
+   TR::java_util_Vector_subList,
+   TR::java_util_concurrent_ConcurrentHashMap_addCount,
+   TR::java_util_concurrent_ConcurrentHashMap_tryPresize,
+   TR::java_util_concurrent_ConcurrentHashMap_transfer,
+   TR::java_util_concurrent_ConcurrentHashMap_fullAddCount,
+   TR::java_util_concurrent_ConcurrentHashMap_helpTransfer,
+   TR::java_util_concurrent_ConcurrentHashMap_initTable,
+   TR::java_util_concurrent_ConcurrentHashMap_tabAt,
+   TR::java_util_concurrent_ConcurrentHashMap_casTabAt,
+   TR::java_util_concurrent_ConcurrentHashMap_setTabAt,
+   TR::java_util_concurrent_ConcurrentHashMap_TreeBin_lockRoot,
+   TR::java_util_concurrent_ConcurrentHashMap_TreeBin_contendedLock,
+   TR::java_util_concurrent_ConcurrentHashMap_TreeBin_find,
+   TR::java_util_TreeMap_all,
+   TR::java_util_EnumMap_put,
+   TR::java_util_EnumMap_typeCheck,
+   TR::java_util_EnumMap__init_,
+   TR::java_util_EnumMap__nec_,
+   TR::java_util_HashMap_all,
+   TR::java_util_ArrayList_all,
+   TR::java_util_Hashtable_all,
+   TR::java_util_concurrent_ConcurrentHashMap_all,
+   TR::java_util_Vector_all,
+
+   // The following list is identified after running sanity.functional tests
+   TR::java_util_AbstractCollection_all,
+   TR::java_util_ArrayDeque_all,
+   TR::java_util_ComparableTimSort_all,
```

Comment:
I think an array with a null-restricted component type could be passed to methods in `ComparableTimSort` by way of `java/util/Arrays.sort`.  Although I don't think null references will end up being stored into such an array, run-time checking of whether the array is flattened might still be needed.

The same is likely true of `TimSort` below.

Reply:
Because `ComparableTimSort` or `TimSort` works on the existing array, if the array has null restricted array component type, there should be no array element with null value. So it is safe to copy between the array elements and no null value store check is required.

Reply:
Right!  I was getting myself confused about the circumstances where the call to`storeFlattenableArrayElementNonHelper` is suppressed.  I was thinking the call would be suppressed in `Walker.cpp` if `safeToSkipNonNullableArrayNullStoreCheck` is true, even if flattening is enabled.  Sorry for my confusion.